[GStreamer] _gst_util_uint64_scale_int: assertion "num >= 0" failed

crmurillo opened the pull request:
#### 2a29f6f8ced1411af0a85aa28ab15a8dd6f2551c
<pre>
[GStreamer] _gst_util_uint64_scale_int: assertion &quot;num &gt;= 0&quot; failed
<a href="https://bugs.webkit.org/show_bug.cgi?id=299529">https://bugs.webkit.org/show_bug.cgi?id=299529</a>

Reviewed by Philippe Normand.

Refactor the calculation of the display-aspect-ratio.

The calculation of the display width and height overflows if the values for the pixel-aspect-ratio numerator and denominator are too large.

This patch refactors the code to separate the calculation of the display-aspect-ratio into a separate and testable function.
The calculation for the display aspect ratio is replaced by gst_video_calculate_display_ratio assumming a regular display (1:1), maintaining the previous behavior.

A unittest has been added to verify the calculation with different sets of width, height and pixel-aspect-ratio.
A layout test has been added to test the behavior when a video has a large pixel-aspect-ratio.

The layout test video-large-pixel-aspect-ratio.html is disabled for the mac platform. The test fails because the returned values for the width and the height are zero.

Test: Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp
* LayoutTests/media/content/video-with-large-par.mp4: Added.
* LayoutTests/media/video-large-pixel-aspect-ratio-expected.txt: Added.
* LayoutTests/media/video-large-pixel-aspect-ratio.html: Added.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getDisplaySize):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateVideoSizeAndOrientationFromCaps):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
(TestWebKitAPI::TEST_F(GStreamerTest, displayAspectRatioCalculation)):

Canonical link: <a href="https://commits.webkit.org/301767@main">https://commits.webkit.org/301767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85e6fc212d6ccc01f6d73eb7e24bfd594b6caeed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126806 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133754 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78421 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54983 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96492 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64523 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129754 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37664 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113429 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77012 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36541 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31606 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77148 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107487 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31911 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136334 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53479 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105003 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53975 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109788 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104707 "Found 1 new API test failure: WebKitGTK/TestWebKitWebXR:/webkit/WebKitWebXR/permission-request (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26734 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50210 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28554 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50942 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53408 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59210 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52664 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55998 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54412 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -247,6 +247,31 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
     return true;
 }
+
+bool getDisplaySize(WebCore::IntSize originalSize, int pixelAspectRatioNumerator, int pixelAspectRatioDenominator, guint& width, guint& height)
```
We don't use `guint` much anymore, can you make these `unsigned`?

crmurillo replied:
yes, I will do it, if needed, because I will also change the return type like @calvaris suggested.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -247,6 +247,31 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
+
+    // Apply DAR to original video size. This is the same behavior as in xvimagesink's setcaps function.
+    if (!(originalSize.height() % height)) {
+        GST_WARNING("Keeping video original height");
```
That's a GST_DEBUG in the previous code, can you keep it?

crmurillo replied:
I will use GST_DEBUG instead of GST_DEBUG_OBJECT since the pipeline object is not available.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -247,6 +247,31 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
+        width = gst_util_uint64_scale_int(originalSize.height(), width, height);
+        height = originalSize.height();
+    } else if (!(originalSize.width() % width)) {
+        GST_WARNING("Keeping video original width");
```
Ditto

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -247,6 +247,31 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
+        height = gst_util_uint64_scale_int(originalSize.width(), height, width);
+        width = originalSize.width();
+    } else {
+        GST_WARNING("Approximating while keeping original video height");
```
Ditto

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3813,32 +3813,14 @@ void MediaPlayerPrivateGStreamer::updateVideoSizeAndOrientationFromCaps(const Gs
+    guint width = 0, height = 0;
+
+    if (!getDisplaySize(originalSize, pixelAspectRatioNumerator, pixelAspectRatioDenominator, width, height)) {
+        GST_WARNING("Failed to get the size after applying the display aspect-ratio");
```
GST_WARNING_OBJECT

philn commented on Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
```diff
@@ -221,6 +222,48 @@ TEST_F(GStreamerTest, capsFromCodecString)
+TEST_F(GStreamerTest, displayAspectRatioCalculation)
+{
+#define TEST_DAR_CALCULATION(videoWidth, videoHeight, parN, parD, displayWidth, displayHeight) G_STMT_START { \
+        GstCaps* caps = gst_caps_new_simple("video/x-raw", "width", G_TYPE_INT, videoWidth, "height", G_TYPE_INT, videoHeight, "pixel-aspect-ratio", GST_TYPE_FRACTION, parN, parD, nullptr); \
```
`auto caps = adoptGRef(...)` and then `caps.get()` when needed.

philn commented on Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
```diff
@@ -221,6 +222,48 @@ TEST_F(GStreamerTest, capsFromCodecString)
+    TEST_DAR_CALCULATION(720, 576, 16, 15, 768, 576); // PAL 4:3
+    TEST_DAR_CALCULATION(352, 288, 12, 11, 384, 288); // CIF 352x288 (≈4:3)
+    TEST_DAR_CALCULATION(1920, 1080, 4, 3, 2560, 1080); // Anamorphic HD (64:27)
+    TEST_DAR_CALCULATION(720, 576, 16, 15, 768, 576); // PAL 4:3 (repeat)
```
Why is this one duplicated?

crmurillo replied:
The duplicated is not needed, I will change it for another set of test values.

philn commented on Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
```diff
@@ -221,6 +222,48 @@ TEST_F(GStreamerTest, capsFromCodecString)
+        EXPECT_EQ(outParN, parN); \
+        EXPECT_EQ(outParD, parD); \
+        \
+        guint calculatedDisplayWidth = 0, calculatedDisplayHeight = 0; \
```
`unsigned`

crmurillo replied:
this was not needed after changing the return type.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -83,6 +83,7 @@ WARN_UNUSED_RETURN GstPad* webkitGstGhostPadFromStaticTemplate(GstStaticPadTempl
 bool getVideoSizeAndFormatFromCaps(const GstCaps*, WebCore::IntSize&, GstVideoFormat&, int& pixelAspectRatioNumerator, int& pixelAspectRatioDenominator, int& stride, double& frameRate, PlatformVideoColorSpace&);
 std::optional<FloatSize> getVideoResolutionFromCaps(const GstCaps*);
 bool getSampleVideoInfo(GstSample*, GstVideoInfo&);
+bool getDisplaySize(WebCore::IntSize originalSize, int pixelAspectRatioNumerator, int pixelAspectRatioDenominator, guint& width, guint& height);
```
So we are returning another size, meaning width and height. We should do it with another IntSize object instead of output parameters. As we need to know if the conversion was successful, we should wrap it inside a std::optional. Something like:

```
std::optional<WebCore::IntSize> computeDisplaySize(WebCore::IntSize, int pixelAspectRatioNumerator, int pixelAspectRatioDenominator);
```

crmurillo replied:
yes, that makes sense

crmurillo committed:
[GStreamer] _gst_util_uint64_scale_int: assertion "num >= 0" failed
https://bugs.webkit.org/show_bug.cgi?id=299529

Reviewed by Philippe Normand.

Refactor the calculation of the display-aspect-ratio.

The calculation of the display width and height overflows if the values for the pixel-aspect-ratio numerator and denominator are too large.

This patch refactors the code to separate the calculation of the display-aspect-ratio into a separate and testable function.
The calculation for the display aspect ratio is replaced by gst_video_calculate_display_ratio assumming a regular display (1:1), maintaining the previous behavior.

A unittest has been added to verify the calculation with different sets of width, height and pixel-aspect-ratio.
A layout test has been added to test the behavior when a video has a large pixel-aspect-ratio.

The layout test video-large-pixel-aspect-ratio.html is disabled for the mac platform. The test fails because the returned values for the width and the height are zero.

Test: Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp
* LayoutTests/media/content/video-with-large-par.mp4: Added.
* LayoutTests/media/video-large-pixel-aspect-ratio-expected.txt: Added.
* LayoutTests/media/video-large-pixel-aspect-ratio.html: Added.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getDisplaySize):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateVideoSizeAndOrientationFromCaps):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
(TestWebKitAPI::TEST_F(GStreamerTest, displayAspectRatioCalculation)):

Canonical link: https://commits.webkit.org/301767@main
